add support to switch column types in data import

javierluraschi opened the pull request:
This pull requests adds support to change data types and include these changes in the generated r-script for data import.

<img width="1440" alt="screen shot 2016-01-21 at 12 41 32 pm" src="https://cloud.githubusercontent.com/assets/3478847/12493904/5806938e-c03c-11e5-9bda-3122e44310bf.png">

jjallaire commented:
I think we might want a bit more vertical space between the column name and the column type dropdown.

jjallaire commented:
I think we should wrap the code preview so as to whenever possible prevent horizontal scrolling. R will actually do a nice job of this for you if you parse your expression then call deparse on it along w/ the width.cutoff argument.

jjallaire commented:
@hadley Any feedback on the layout of controls, nomenclature used, etc.?

jmcphers commented on src/cpp/session/resources/grid/gridstyles.css:
```diff
@@ -317,3 +317,15 @@ div.DTS div.DTS_Loading.showLoading {
+.columnClickable {
+    cursor: pointer;
+    background: url("datatables/images/header_expand.png") no-repeat right top;
+    padding-right: 15px !important;
```
We generally try to make a rule more specific to create an overriding style (since `!important` is itself difficult to override), but this is okay if there's no other straightforward way to apply the style.

javierluraschi replied:
Fixed.

jmcphers commented on src/gwt/src/org/rstudio/core/client/widget/GridViewerFrame.java:
```diff
@@ -50,4 +62,17 @@ private final native void setDataNative(WindowEx frameContentWindow, JavaScriptO
       frameContentWindow.setOption(option, value);
    }-*/;
+   
+   private final native void setColumnDefinitionsUIVisibleNative(WindowEx frameContentWindow, boolean value, Operation onColumnDefsChange) /*-{
```
We generally try to keep code to 80 columns wide -- unfortunately there's no way to make Eclipse do this automatically, but you can make it show a print margin at 80 columns so you know where to hard-wrap.

javierluraschi replied:
Will try on a best effort basic, fixed this one.

jmcphers commented:
More UI feedback:
- as @jjallaire mentioned we want to avoid horizontal scrolling in the code preview (it's very likely we'll have vertical scrolling and a textarea that scrolls both ways is hard to use); perhaps we could also enable soft wrapping in Ace? 
- I'd add a small amount of breathing room around the code in the preview so the characters aren't right up against the edges
- Many users interpret gray text as disabled content, so that treatment may not make the data types appear to be editable. The column headers are already bolded so probably just leaving these in regular type (and perhaps at -1px in size) is enough differentiation; if we're worried that the column name and type will be conflated we could add some more padding and/or a subtle line between them.

jmcphers commented on src/cpp/session/resources/grid/gridviewer.js:
```diff
@@ -611,6 +624,83 @@ var createFilterUI = function(idx, col) {
+var openColumnTypeChooserUI = function(idx, col, ele, onDismiss) {
+  invokeFilterPopup(ele, function(popup) {
+    var columnTypes = [
+      "logical", "integer", "double", "character", "date", "datetime", "numeric" 
```
looks like it's missing some column types -- should we include `factor`, `time`, etc?

javierluraschi replied:
I based the supported types from `vignette("column-types")`; however, those seem only the auto-detectable ones. I'll add all the types I find under `readr::col_*`

javierluraschi replied:
Fixed.

jmcphers commented on src/cpp/session/modules/SessionDataViewer.R:
```diff
@@ -91,6 +92,7 @@
     if (length(x[[idx]]) > 0 && length(x[[idx]][1]) == 1)
     {
       val <- x[[idx]][1]
+      col_type_r <- typeof(val)
```
We should discuss this with @hadley. I think that several different column types from the perspective of `readr` may map to what appears to be the same type in R (for instance, the `col_number` type parses into vectors of type `double`). That implies that we might need to change the UX a little to show "parse as" in addition to or instead of the inferred data type of the result.

javierluraschi replied:
Hadley took a look, no comments here so I'll keep it as-is for now.

hadley commented:
A few thoughts/questions:

* How are you planning on handling parsers that have additional parameters - for example, for `col_datetime()` you might want to provide your own date formatting string, or for `col_factor()` you have to supply a list of factors.

* I'd collapse the two escaping check marks into a drop down box (none, backslash, double, both).  Be aware that the defaults for `read_csv()` and `read_delim()` are different.

* Columns names -> Column names. But might be better as "First row is column names?"

* I think you might want to do a bit more munging to create the default variable name. I'd at least lower case it.

* How many rows are you parsing by default? Have you thought at all about how you might surface the results of parsing problems? Maybe: "Previewing first 50 entries: 34 problems"  It might be nice to have a row at the bottom that gives the number of parsing errors for the first x rows.

* What happens if a cell contains a very long string or multiple newlines?

* What happens if the encoding isn't as UTF-8 and you get undisplayable characters?

* You should definitely add "skip" as a parsing option, and then grey out that column

* It might be worth using the dplyr abbreviations for column types to avoid making narrow cols wide.

javierluraschi commented:
@hadley do you have feedback for the locale drop down as well? In this current iteration, we get all the locales from `readr::date_names_langs()`, push them to the locale dropdown and generate a parameter as `locale(date_names=drop-down-value)`. However, we currently don't support the other ones: date_format, time_format, decimal_mark = ".", tz = "UTC", encoding = "UTF-8" and asciify. It would also help me to understand what is the main usage of `locale` for users of `readr::` to consider adding additional entries or mechanisms to set/use user preferences based on their selection. It might make sense to have a configuration button that opens a dialog to set all these parameters once and save them as user preferences across imports. Thoughts?

hadley commented:
@javierluraschi I don't have a good feel for how the locales should work - the way I imagine people using this in practice is to develop a set of options that work for their data and then save and reuse it across projects. Exposing that in the UI would be very nice! 

Language is probably the least important setting because it's only needed when parsing dates and times (and then only for month names and am/pm equivalents). Decimal mark is the most important and problem should be hoisted to the top level. Encoding is probably next most important, because it will affect any strings. TZ is also important, although it's typically something people set after they've discovered the default is wrong.

javierluraschi commented:
@jjallaire 
- Vertical space added.
- Preventing horizontal scroll with deparse.
- Improved on Hadleys layout feedback.

@jmcphers 
- Some comments inline, fixed now.
- Using soft wrapping in ace.
- Added breathing room in code editor.
- Make headers black, they do look better now.

@hadley 
- `col_datetime`, `col_factor`, etc. now prompt the user for a format string. We might be able to improve a bit more at some point, but this approach looks reasonable to me. Will be happy to provide more ad-hoc UI if we feel strongly about this.
- Collapsed escaping options into dropdown.
- "Column names" thanks for finding this typo!
- Improved variable names, underscores, lowercase, etc.
- Parsing errors not included in this PR, added an idonethis todo do consider this later today.
- Long strings and new lines in input not considered yet, added an idonethis todo to consider this further
- Test and address undisplayable characters, added idonethig todo.
- "Skip" option added with column greying. Added support to "Include" a column (default), "Skip" or "Only" import the selected columns.
- Thank you so much for the locales feedback. Makes sense. However, since this PR is getting really big, I've added an idonethis todo to follow up with a subsequent one.

<img width="1440" alt="screen shot 2016-01-25 at 3 08 55 pm" src="https://cloud.githubusercontent.com/assets/3478847/12568065/e5366c28-c377-11e5-9311-7f73212a15dc.png">

hadley commented:
@javierluraschi Looking good. Two nit-picky details:

* The import options look a bit weird - I think they might look better visually if they were contained in a well or similar.

* default and none should be Default and None

jmcphers commented:
@javierluraschi @jjallaire changes to `gridviewer.js` look good!

javierluraschi commented:
![Uploading Screen Shot 2016-01-26 at 10.38.51 AM.png…]()

javierluraschi commented:
@jjallaire removing `readr::` namespace
@jmcphers thanks!
@hadley thanks, changed. Regarding options block. How about this? It's matching the color of the data preview header/footer.

<img width="1440" alt="screen shot 2016-01-26 at 10 38 51 am" src="https://cloud.githubusercontent.com/assets/3478847/12590818/a1496bb2-c419-11e5-9257-ec9025bc7b9c.png">
